SmartDeviceLink_Example/ButtonManager.swift

```diff
@@ -0,0 +1,159 @@
+        didSet {
+            guard let refreshUIHandler = refreshUIHandler else { return }
+            refreshUIHandler()
+        }
+    }
+
+    /// SDL UI images are visible if true; hidden if false
+    public fileprivate(set) var imagesEnabled: Bool {
+        didSet {
+            guard let refreshUIHandler = refreshUIHandler else { return }
+            updateToggleSoftButtonIcon(isEnabled: toggleEnabled, isImageEnabled: imagesEnabled)
+            updateAlertSoftButtonIcon()
+            refreshUIHandler()
+        }
+    }
+
+    /// Keeps track of the toggle soft button current state. The image or text changes when the button is selected
+    fileprivate var toggleEnabled: Bool {
+        didSet {
+            updateToggleSoftButtonIcon(isEnabled: toggleEnabled, isImageEnabled: imagesEnabled)
+        }
+    }
+
+    init(sdlManager: SDLManager, updateScreenHandler: refreshUIHandler? = nil) {
+        self.sdlManager = sdlManager
+        self.refreshUIHandler = updateScreenHandler
+        textEnabled = true
+        imagesEnabled = true
+        toggleEnabled = true
+        super.init()
+    }
+
+    /// Creates and returns an array of all soft buttons for the UI
+    ///
+    /// - Parameter manager: The SDL Manager
+    /// - Returns: An array of all soft buttons for the UI
+    func allScreenSoftButtons(with manager: SDLManager) -> [SDLSoftButtonObject] {
+        return [softButtonAlert(with: manager), softButtonToggle(), softButtonTextVisible(), softButtonImagesVisible()]
+    }
+}
+
+// MARK: - Custom Soft Buttons
+
+private extension ButtonManager {
+    /// Returns a soft button that shows an alert when tapped.
+    ///
+    /// - Parameter manager: The SDL Manager for showing the alert
+    /// - Returns: A soft button
+    func softButtonAlert(with manager: SDLManager) -> SDLSoftButtonObject {
+        let imageSoftButtonState = SDLSoftButtonState(stateName: AlertSoftButtonImageState, text: nil, image: UIImage(named: CarIconImageName))
+        let textSoftButtonState = SDLSoftButtonState(stateName: AlertSoftButtonTextState, text: AlertSoftButtonText, image: nil)
+        let softButton: SDLSoftButtonObject? = SDLSoftButtonObject(name: AlertSoftButton, states: [imageSoftButtonState, textSoftButtonState], initialStateName: imageSoftButtonState.name) { (buttonPress, buttonEvent) in
+            guard buttonPress != nil else { return }
+            let alert = AlertManager.alertWithMessageAndCloseButton("You pressed the button!")
+            manager.send(alert)
+        }
+
+        guard softButton != nil else { assert(false, "Button was not created successfully") }
+        return softButton!
+    }
+
+    /// Returns a soft button that toggles between two states: on and off. If images are currently visible, the button image toggles; if images aren't visible, the button text toggles.
+    ///
+    /// - Returns: A soft button
+    func softButtonToggle() -> SDLSoftButtonObject {
```

Comment:
I think we should remove the full image / text toggling soft buttons. It isn't a use case people will be looking to an example for, and it's become very complicated (and will probably confuse people and make SDL appear more complicated than it is). A toggling soft button that changes a soft button when pressed is probably sufficient.

Reply:
Fixed. Toggle button now switches between two images.